src/StreamUtils.bsv

```diff
@@ -0,0 +1,376 @@
+    
+    rule prepareStream;
+        let stream = inputFifo.first;
+        inputFifo.deq;
+        StreamWithPtr streamWithPtr = StreamWithPtr {
+            stream: stream,
+            bytePtr: convertByteEn2BytePtr(stream.byteEn) 
+        };
+        prepareFifo.enq(streamWithPtr);
+    endrule
+
+    rule assertSplitStream;
+        let stream = prepareFifo.first.stream;
+        let bytePtr = prepareFifo.first.bytePtr;        
+        let splitLocation = splitLocationFifo.first;
+        DataBytePtr truncateBytePtr = 0;
+        if (!isSplitted && unpack(zeroExtend(bytePtr)) + streamByteCntReg >= splitLocation) begin
+            truncateBytePtr = truncate(pack(splitLocation - streamByteCntReg));
+        end
+        DataBytePtr resBytePtr = getMaxBytePtr() - truncateBytePtr;
+        splitPtrFifo.enq(tuple2(truncateBytePtr, resBytePtr));
+        if (truncateBytePtr > 0 && !stream.isLast) begin
+            isSplitted <= True;
+        end 
+        else if (stream.isLast) begin
+            isSplitted <= False;
+        end
+        streamByteCntReg <= stream.isLast ? 0 : streamByteCntReg + unpack(zeroExtend(bytePtr));
+        assertFifo.enq(prepareFifo.first);
+        prepareFifo.deq;
+        if (stream.isLast) begin
+            splitLocationFifo.deq;
+        end
+    endrule
+
+
+    rule execSplitStream;
+        // Only output remainStreamReg
+        if (hasRemainReg && hasLastRemainReg) begin
+            outputFifo.enq(remainStreamReg);
+            hasRemainReg <= False;
+            hasLastRemainReg <= False;
+        end
+        else if (assertFifo.notEmpty && splitPtrFifo.notEmpty) begin
+            let stream = assertFifo.first.stream;
+            let frameBytePtr = assertFifo.first.bytePtr;
+            let {truncateBytePtr, resBytePtr} = splitPtrFifo.first;
+            assertFifo.deq;
+            splitPtrFifo.deq;
+            // no operatation
+            if (!hasRemainReg && truncateBytePtr == 0) begin
+                outputFifo.enq(stream);
+            end
+            // split the frame in this cycle to a last frame and a remain frame
+            else if (!hasRemainReg && truncateBytePtr > 0) begin
+                DataBitPtr truncateBitPtr = zeroExtend(truncateBytePtr) << valueOf(BYTE_WIDTH_WIDTH);
+                DataBitPtr resBitPtr = zeroExtend(resBytePtr) << valueOf(BYTE_WIDTH_WIDTH);
+                outputFifo.enq(DataStream {
+                    data: (stream.data << resBitPtr) >> resBitPtr,
+                    byteEn: (stream.byteEn << resBytePtr) >> resBytePtr,
+                    isFirst: stream.isFirst,
+                    isLast: True
+                });
+                DataStream remainStream = DataStream {
+                    data: stream.data >> truncateBitPtr,
+                    byteEn: stream.byteEn >> truncateBytePtr,
+                    isFirst: True,
+                    isLast: True
+                };
+                hasRemainReg <= unpack(remainStream.byteEn[0]);
```

Comment:
very poor readability.

first, for a new person to read code, it won't easy to know that byteEN has continous ones and the only lsb check is enough.

second,  even for a person who is familiar with the code, he still need to pause and think that you are using the lsb trick to do the check.

make your code be like an article, for example, if you define a function called `isByteEnAllZero`, then you code will be 
```
hasRemainReg <= isByteEnAllZero(remainStream.byteEn);
```
which read like an `English sentence`, not a `code`, and for the one who reads the code, he doesn't need any additional effort to figure out how does this function work, he can focus his thought in the current code, and no need to jump to somewhere else to know the underlying logic.

Good code doesn't need comment, since the code itself is an article that tells you how does the code works.

third，there are more than one place that use this logic.

Form the point of software enginering, this is encapsulation, abstraction, layering, and reuse。